### PR TITLE
Assignment2: Tyler + Austin

### PR DESCRIPTION
**Pull request by twoody0**
Partnered with @AustinH5544

### CODE REVIEW

File: Logger/LogFactory.cs [hunk 1]
```diff
@@ -1,10 +1,35 @@
...
     {
+        if (string.IsNullOrEmpty(FilePath))
+        {   
+            return null;
+        }
+        return new FileLogger(FilePath) { ClassName = className };
+    }
 
-        return null;
+    private static string NormalizePath(string path)
```
Comment (theRoice) [inline Logger/LogFactory.cs:29]: Normalizing the path is is a great idea, we might have to steal this.

**Comment by theRoice** [comment]
- There is an existing `BaseLogger` class. It needs an **auto property** to hold a class name. This property should be set in the `LogFactory` using an **object initializer**. ✔
- Create a `FileLogger` that derives from `BaseLogger`. It should take in a path to a file to write the log message to. When its `Log` method is called, it should **append** messages on their own line in the file. The output should include all of the following:
  - The current date/time ✔
  - The name of the class that created the logger ✔
  - The log level ✔
  - The message ✔
  - The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
- The `LogFactory` should be updated with a new method `ConfigureFileLogger`. This should take in a file path and store it in a **private member**. It should use this when instantiating a new `FileLogger` in its `CreateLogger` method. ✔
- If the file logger has not be configured in the `LogFactory`, its `CreateLogger` method should return `null`. ✔
- Inside of `BaseLoggerMixins` implement **extension methods** on `BaseLogger` for
  - `Error`, ✔
  - `Warning`, ✔
  - `Information`, and ✔
  - `Debug`. ✔
  Each of these methods should take in a `string` for the message, as well as a **parameter array** of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the `BaseLogger.Log` method, by automatically supplying the appropriate `LogLevel`. These methods should throw an exception if the `BaseLogger` parameter is null. There are a couple example unit tests to get you started.
- Use the nameof() operator when identifying the class name to the logger ✔
- Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
- Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
- Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
- Ensure that you turn on EnforceCodeStyleInBuild ✔
- Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional) ✔
- Turn on Nullability (`Nullable`) ✔
- **All of the above should be unit tested.**

## Extra Credit

- Implement an additional logger. This logger must be unit tested. Some options to consider could be one that uses `System.Console` or `System.Diagnostics.Trace` ✔
- Implement the factory pattern using static abstract methods on a Logger interface instead.❌

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 2]
```diff
@@ -1,47 +1,105 @@
...
 
-}
-
-public class TestLogger : BaseLogger
-{
-    public List<(LogLevel LogLevel, string Message)> LoggedMessages { get; } = new List<(LogLevel, string)>();
 
-    public override void Log(LogLevel logLevel, string message)
+    [TestMethod]
+    public void ClassName_TestClass_Success()
```
Comment (theRoice) [inline Logger.Tests/BaseLoggerMixinsTests.cs:93]: Consider changing this name to better represent what this test does and expanding it using data rows to test some weird strings too. Maybe some variation of Logger_GivenClassNameString_UpdatesSuccessfully()?

File: Logger.Tests/FileLoggerTests.cs [hunk 3]
```diff
@@ -1,9 +1,82 @@
...
+        //Act
+        FileLogger logger = new(FilePath);
+        logger.ClassName = className;
+
+        //Assert
+        Assert.AreEqual(className, logger.ClassName);
+    }
+
+    [TestMethod]
+    public void FileLogger_Constructor_Success()
```
Comment (theRoice) [inline Logger.Tests/FileLoggerTests.cs:36]: I'd consider looking at the tests in this file and updating the test method names just a little bit. For example, on this line, you might consider "FileLogger_OnInstantiation_IsNotNull()". This lets us know what object/method is being called, the conditions under which we are testing, and the result we are expecting.

File: Logger.Tests/FileLoggerTests.cs [hunk 4]
```diff
@@ -1,9 +1,82 @@
...
+        logger.Log(logLevel, message);
+
+        //Assert
+        string logContents = File.ReadAllText(FilePath);
+        string expectedLogEntry = $"{className} {logLevel}: {message}";
+        Assert.IsTrue(logContents.Contains(expectedLogEntry));
+    }
 
+    [TestCleanup]
+    public void Cleanup()
```
Comment (theRoice) [inline Logger.Tests/FileLoggerTests.cs:76]: Good idea with the cleanup! We might end up doing this!

File: Logger.Tests/Logger.Tests.csproj [hunk 5]
```diff
@@ -1,9 +1,15 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
 	<PropertyGroup>
+    <AnalysisLevel>latest-Recommended</AnalysisLevel>
 		<TargetFramework>net7.0</TargetFramework>
```
Comment (theRoice) [inline Logger.Tests/Logger.Tests.csproj:5]: net8.0 is the current LTS version

Comment (twoody0) [reply Logger.Tests/Logger.Tests.csproj:5]: Thank you! I realized that while I was looking at other people's code that both mien and theirs wasn't fully up to date

File: Logger.Tests/FileLoggerTests.cs (hunk 4 above)
Comment (twoody0) [reply Logger.Tests/FileLoggerTests.cs:76]: Feel free to! :)

File: Logger.Tests/BaseLoggerMixinsTests.cs (hunk 2 above)
Comment (twoody0) [reply Logger.Tests/BaseLoggerMixinsTests.cs:93]: Great idea! That would definitely help understand what the test is doing.

File: Logger/LogFactory.cs (hunk 1 above)
Comment (twoody0) [reply Logger/LogFactory.cs:29]: Go ahead!

File: Logger/BaseLogger.cs [hunk 6]
```diff
@@ -2,6 +2,7 @@
 
 public abstract class BaseLogger
 {
+    public string ClassName { get; set; } = string.Empty;
```
Comment (LeoMarlow) [inline Logger/BaseLogger.cs:5]: Hey boys pretty sure this line here is a problem.  we use "public string? ClassName { get; set; }"to fix this when setting the string to an empty string I believe you are removing the auto name property of baselogger to be Logfactory.

**Review by LeoMarlow — COMMENTED** [review]
Good job boys I added a single comment on the line I thought was a problem but other than that it looks pretty clean. 
Assignment
There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer.  ?✔?
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ✔
The log level ✔
The message ✔
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
Use the nameof() operator when identifying the class name to the logger ✔
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ✔

File: Logger/BaseLogger.cs (hunk 6 above)
Comment (twoody0) [reply Logger/BaseLogger.cs]: Good catch, we originally had className be a field, so we needed it to be set to string.Empty, but we forgot to remove that after we turned it into a property. Thank you!

File: Logger.Tests/FileLoggerTests.cs (hunk 3 above)
Comment (twoody0) [reply Logger.Tests/FileLoggerTests.cs]: Thanks for the advice! We renamed the tests per your suggestion

File: Logger/TraceLogger.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
+﻿using System.Diagnostics;
+using System;
+using System.IO;
+
+namespace Logger;
+
+public class TraceLogger : BaseLogger
+{
+    private readonly FileLogger _fileLogger;
+    public TraceLogger(string className, FileLogger fileLogger)
```
Comment (BenjaminMichaelis) [inline Logger/TraceLogger.cs:10]: Why does `TraceLogger` take in a `FileLogger`? This seems strange.... I would not expect a Trace logger to also write the file.

**Pull request closed without merging**